File: 2021/12-Passage-Pathing.py
import sys

adjacent = {}

for line in sys.stdin:
    a,b = line.replace('\n', '').split('-')
    
    if not a in adjacent:
        adjacent[a] = []
    if not b in adjacent:
        adjacent[b] = []

    adjacent[a].append(b)
    adjacent[b].append(a)

def find_paths(start, end, paths=[], currentPath=[], visited = set()):
    if start == end:
        paths.append(currentPath + [end])
        return paths

    if start in visited:
        return paths

    my_visited = visited | {start} if start.islower() else visited
    my_path = currentPath + [start]

    for next in adjacent[start]:
        find_paths(next, end, paths, my_path, my_visited)

    return paths

def find_paths2(start, end, paths=[], currentPath=[], visited = set(), double=False):
    if start == end:
        paths.append(currentPath + [end])
        return paths

    if start in visited and (start == 'start' or double):
        return paths

    my_visited = visited
    if start.islower():
        if start in visited:
            double = True
        else:
            my_visited = visited | {start}

    my_path = currentPath + [start]

    for next in adjacent[start]:
        find_paths2(next, end, paths, my_path, my_visited, double)

    return paths

def part_one():
    paths = find_paths('start', 'end')

    print(len(paths))

def part_two():
    paths = find_paths2('start', 'end')

    print(len(paths))

print(part_one())
print(part_two())

File: 2021/12-Passage-Pathing.js
const readline = require('readline')
const rl = readline.createInterface({ input: process.stdin })

const nodes = new Set()
const graph = []

const adjacent = new Map()

const visitOnce = new Set()

function addEdge(a,b) {
  if (!adjacent.has(a)) {
    adjacent.set(a,new Set())
  }
  if(!adjacent.has(b)) {
    adjacent.set(b,new Set())
  }

  adjacent.get(a).add(b)
  adjacent.get(b).add(a)

  if (a.toLowerCase() === a)
    visitOnce.add(a)
  if (b.toLowerCase() === b)
    visitOnce.add(b)
}

rl.on("line", data => {
  const edge = data.split('-')
  addEdge(...edge)
  nodes.add(edge[0])
  nodes.add(edge[1])
})

rl.on('close', () => {
  console.log(adjacent)
  console.log(visitOnce)
  
  console.log('partOne', partOne(graph))
  console.log('partTwo', partTwo(graph))
})

function findPaths(from, to) {
  return solve({from, to})

  function solve({from, to, paths=[], currentPath=[], visited=new Set()}) {
    if (from === to) {
      paths.push([...currentPath, to])
      return paths
    }

    if (visited.has(from)) return paths

    const myVisited = new Set([...visited])

    if (visitOnce.has(from)) myVisited.add(from)

    for (const next of adjacent.get(from)) {
      solve({from:next, to, paths, currentPath:[...currentPath, from], visited:myVisited})
    }
    return paths
  }
}

function findPaths2(from, to) {
  return solve({from, to})

  function solve({from, to, paths=[], currentPath=[], visited=new Set(), hasDouble=false}) {
    if (from === to) {
      paths.push([...currentPath, to])
      return paths
    }

    if (visited.has(from) && (from === 'start' || hasDouble)) return paths

    const myVisited = new Set([...visited])

    if (visitOnce.has(from)) {
      if (myVisited.has(from))
        hasDouble = true
      else
        myVisited.add(from) 
    }

    for (const next of adjacent.get(from)) {
      solve({from:next, to, paths, currentPath:[...currentPath, from], visited:myVisited, hasDouble})
    }
    return paths
  }
}



function partOne() {
  const paths = findPaths('start', 'end')

  return paths.length
}

function partTwo() {
  const paths = findPaths2('start', 'end')

  return paths.length
}
